arch/xtensa/src/common/xtensa_coproc.S

```diff
@@ -89,31 +85,18 @@ _xtensa_coproc_saoffsets:
  *   around the assembly language call to _xtensa_coproc_savestate.
  *
  * Entry Conditions:
- *   - A2 holds the address of current interrupt stack pointer.
+ *   in - Pointer to where the save is performed (e.g. a2)
  *   - Registers have been saved/spilled already.
  *
  * Exit conditions:
  *   - All necessary CP callee-saved state has been saved.
- *   - Registers a2-a7, a13-a15 have been trashed.
- *
- * Must be called from assembly code only, using CALL0.
+ *   - Registers a2-a7, a13-a14 have been trashed.
  *
  ****************************************************************************/
 
-  .global	_xtensa_coproc_savestate
-  .type	_xtensa_coproc_savestate, @function
-
-  .align	4
-  .literal_position
-  .align	4
-
-_xtensa_coproc_savestate:
-
-	/* Move the address of the thread state save area to R15 */
+	.macro xtensa_coproc_savestate in
```

Comment:
why need in argument? Since xtensa_coproc_savestate already touch many registers(a2-a7, a13-a14), we can directly:
```
.macro xtensa_coproc_savestate
add a3, sp, (4 * XCPTCONTEXT_REGS)
```
and keep the rest unchanged.